tiledb/sm/filter/bitsort_filter.cc

```diff
@@ -0,0 +1,450 @@
+      const Tile& tile,
+      Tile* const tile_offsets,
+      FilterBuffer* input_metadata,
+      FilterBuffer* input,
+      FilterBuffer* output_metadata,
+      FilterBuffer* output,
+      const Config& config) const {
+  (void)tile;
+  (void)tile_offsets;
+  (void)input_metadata;
+  (void)input;
+  (void)output_metadata;
+  (void)output;
+  (void)config;
+  return Status_FilterError("BitSortFilter: Do not call default version of of run_reverse.");
+}
+
+Status BitSortFilter::run_forward(
+    const Tile& tile,
+    std::vector<Tile*> &dim_tiles,
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+
+  // Since run_forward interprets the filter's data as integers, we case on
+  // the size of the type and pass in the corresponding integer type into
+  // a templated function.
+  auto tile_type = tile.type();
+  switch (datatype_size(tile_type)) {
+    case sizeof(uint8_t): {
+      return run_forward<uint8_t>(
+          dim_tiles, input_metadata, input, output_metadata, output);
+    }
+    case sizeof(uint16_t): {
+      return run_forward<uint16_t>(
+         dim_tiles, input_metadata, input, output_metadata, output);
+    }
+    case sizeof(uint32_t): {
+      return run_forward<uint32_t>(
+          dim_tiles, input_metadata, input, output_metadata, output);
+    }
+    case sizeof(uint64_t): {
+      return run_forward<uint64_t>(
+          dim_tiles, input_metadata, input, output_metadata, output);
+    }
+    default: {
+      return Status_FilterError(
+          "BitSortFilter::run_forward: datatype does not have an appropriate "
+          "size");
+    }
+  }
+}
+
+template <typename AttrType>
+Status BitSortFilter::run_forward(
+    std::vector<Tile*> &dim_tiles,
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  // Output size does not change with this filter.
+  RETURN_NOT_OK(output->prepend_buffer(input->size()));
+  Buffer* output_buf = output->buffer_ptr(0);
+  assert(output_buf != nullptr);
+
+  // Write the metadata.
+  auto parts = input->buffers();
+  auto num_parts = (uint32_t)parts.size();
+  uint32_t metadata_size = sizeof(uint32_t) + num_parts * sizeof(uint32_t);
+  RETURN_NOT_OK(output_metadata->append_view(input_metadata));
+  RETURN_NOT_OK(output_metadata->prepend_buffer(metadata_size));
+  RETURN_NOT_OK(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  // Determine the total size (in bytes) of the input data.
+  size_t total_size = 0;
+  for (uint64_t i = 0; i < num_parts; ++i) {
+    const auto &part = parts[i];
+    total_size += part.size();
+  }
+  size_t elements_size = total_size / sizeof(AttrType);
+
+  // Create a vector to store the attribute data with their respective positions.
+  // Then for each part of the input, collect this data into the sorted_elements vector.
+  std::vector<AttrType> elements;
```

Comment:
Creating this elements vector will be a performance hit and I don't think it's necessary at all... To me it looks like elements will contain the same exact data as in part? If so let's get rid of it and at the same time collect_part_data becomes unnecessary.

Reply:
I mean we could use copy_to. I'll do that for now. But we need a way to read data and I thought this was better than continuously calling FilterBuffer::read. Thoughts?

Reply:
Get a pointer to the data:
`const AttrType* input_data = static_cast<const AttrType*>(parts[i]->data());`

Using that with an index should prevent you from calling FilterBuffer::read.

Reply:
I'd still need to know what i is. Should I keep track of the offsets vector and binary search through it to figure out what i should be? Create a helper function for it?

Reply:
Oh I see, input might be a composition of multiple buffers... Question, do we plan on combining bitsort filter with any other filter types? I'm asking because I'm hoping maybe in this case we might be able to assume there is really only one underlying buffer and we can use that directly to save the copy.